### PR TITLE
Preserve output tokens that equal mask_token_id

### LINKED ISSUE #73: dflash_generate: final mask cleanup is value-based and drops legitimate output tokens equal to mask_token_id
### Summary

`dflash_generate`'s final cleanup at `dflash/model.py:150-151` removes any output token that happens to equal `mask_token_id`, regardless of whether it was generated by the model or just leftover unfilled buffer.

```python
output_ids = output_ids[:, :max_length]
output_ids = output_ids[:, output_ids[0] != mask_token_id]
```

The intent is to drop the trailing mask-padded positions left over when generation stops early. The actual effect is value-based, so any legitimate token whose id equals `mask_token_id` is silently dropped from the middle of the sequence.

### Evidence (no model required)

```python
import torch
mask_token_id = 0
buf = torch.tensor([[11, 12, 13,
                     101, 102, 0, 103, 104, 0, 105, 106, 107, 108,
                     0, 0, 0]])  # last 3 = unfilled mask
stripped = buf[:, buf[0] != mask_token_id]
print(stripped.tolist())
# [[11, 12, 13, 101, 102, 103, 104, 105, 106, 107, 108]]   <- two zeros gone
# expected: [[11, 12, 13, 101, 102, 0, 103, 104, 0, 105, 106, 107, 108]]
```

### Why this matters in practice

The published draft configs choose `mask_token_id` from each tokenizer's reserved/special-token zone (e.g. Llama-3.1: `128002 = <|reserved_special_token_0|>`, Qwen3.5-4B/9B: `248070`, Qwen3-8B-b16: `151669`, gpt-oss-20b: `200000`), which avoids the worst case under the supported HF chat templates. But the cleanup is still fragile by construction:

- A user-supplied or future config that chooses an in-vocabulary id (or `0`) silently truncates output.
- If the model ever emits the special token (e.g. wrong tokenizer revision, raw / non-chat usage), that token vanishes with no error and the reported `num_output_tokens` is wrong.

### Suggested fix

Track the number of valid tokens written during the loop and slice by length, rather than filtering by value. The `start` variable already tracks this:

```python
output_ids = output_ids[:, :start]
# stop-token trim still applied as today
```

If you'd like, I can prepare a small PR with the change plus a unit test that asserts the legitimate-token case.

### Environment

- repo at `HEAD` `6e0c951`
- python 3.12, torch 2.11

### PR DESCRIPTION
**Pull request by shaun0927**
Closes #73.

### Problem

`dflash_generate`'s final cleanup at `dflash/model.py:150-151` strips trailing
mask-padded positions by **value**, not position:

```python
output_ids = output_ids[:, :max_length]
output_ids = output_ids[:, output_ids[0] != mask_token_id]
```

Any legitimate output token whose id equals `mask_token_id` is silently
dropped from the middle of the sequence. The published configs avoid the
worst case by choosing `mask_token_id` from each tokenizer's reserved
zone, but the cleanup is still fragile by construction (a misconfigured
tokenizer or a model that emits the special token loses tokens).

Reproduction without a model:

```python
import torch
mask_token_id = 0
buf = torch.tensor([[11,12,13, 101,102,0,103,104, 0,0,0]])  # last 3 = unfilled
stripped = buf[:, buf[0] != mask_token_id]
# [[11, 12, 13, 101, 102, 103, 104]]   -- the legitimate 0 disappears
```

### Fix

The generation loop already tracks the last written index via `start`
(used at line 139 for `past_key_values_target.crop(start)`). A
length-based slice that caps at `max_length` is equivalent in the normal
case and correct in the pathological case:

```python
output_ids = output_ids[:, :min(start + 1, max_length)]
```

`+ 1` is needed because `output_ids[:, start]` holds the posterior token
sampled in the final iteration (written at index
`old_start + acceptance_length + 1` and then `start` advances to that
index). Capping at `max_length` preserves the existing `max_new_tokens`
guarantee on natural exit.

### Equivalence in the normal case

Numbers from a small mental simulation (`num_input_tokens=5`, `max_length=10`,
`block_size=5`):

| Termination | Original output length | Patched output length |
|---|---|---|
| Natural exit (start overshoots, no in-range mask) | 10 | 10 |
| Early stop, stop token at posterior position | 10 (incl. stop) | 10 (incl. stop) |
| Early stop, stop token mid-sequence | 9 | 9 |
| Natural exit, generated token equals mask_token_id | **9 (bug)** | **10 (correct)** |

A unit-style assertion is included as a comment in `.analysis/verify/k2_fix_test.py`
in the analysis tree (not committed to the repo) to confirm parity on the
first three rows and divergence on the fourth.

### Notes

This is one of three small correctness fixes I'm sending against the
recently merged repo (also #74, #75). Each is on its own branch so they
can be merged or rejected independently.

### CODE REVIEW

**Comment by jianc99** [comment]
Good catch! Thanks.
